pkg/blockbuilder/blockbuilder.go

```diff
@@ -239,3 +238,267 @@ func (b *BlockBuilder) getLagForPartition(ctx context.Context, partition int32)
+		level.Info(logger).Log("msg", "done consuming", "duration", dur, "curr_lag", state.Lag,
+			"last_block_end", startState.LastBlockEnd, "curr_block_end", blockEnd,
+			"last_seen_offset", startState.LastSeenOffset, "curr_seen_offset", state.LastSeenOffset,
+			"num_blocks", numBlocks)
+	}(time.Now(), state)
+
+	// We always rewind the partition's offset to the commit offset by reassigning the partition to the client (this triggers partition assignment).
+	// This is so the cycle started exactly at the commit offset, and not at what was (potentially over-) consumed previously.
+	b.kafkaClient.AddConsumePartitions(map[string]map[int32]kgo.Offset{
+		b.cfg.Kafka.Topic: {
+			partition: kgo.NewOffset().At(state.Commit.At),
+		},
+	})
+	defer b.kafkaClient.RemoveConsumePartitions(map[string][]int32{b.cfg.Kafka.Topic: {partition}})
+
+	level.Info(logger).Log("msg", "start consuming", "offset", state.Commit.At)
+
+	var (
+		firstRec  *kgo.Record
+		lastRec   *kgo.Record
+		commitRec *kgo.Record
+	)
+
+consumerLoop:
+	for remaining := state.Lag; remaining > 0; {
+		if err := context.Cause(ctx); err != nil {
+			return partitionState{}, err
+		}
+		fetches := b.kafkaClient.PollFetches(ctx)
+		fetches.EachError(func(_ string, _ int32, err error) {
+			if !errors.Is(err, context.Canceled) {
+				level.Error(logger).Log("msg", "failed to fetch records", "err", err)
+				b.blockBuilderMetrics.fetchErrors.Inc()
+			}
+		})
+
+		numRecs := fetches.NumRecords()
+		remaining -= int64(numRecs)
+
+		recIter := fetches.RecordIter()
+		for !recIter.Done() {
+			rec := recIter.Next()
+
+			if firstRec == nil {
+				firstRec = rec
+			}
+
+			// Stop consuming after we reached the cycleEnd marker.
+			// NOTE: the timestamp of the record is when the record was produced relative to distributor's time.
+			if rec.Timestamp.After(cycleEnd) {
+				break consumerLoop
+			}
+
+			recProcessedBefore := rec.Offset <= state.LastSeenOffset
+			allSamplesProcessed, err := builder.Process(ctx, rec, state.LastBlockEnd.UnixMilli(), blockEnd.UnixMilli(), recProcessedBefore)
+			if err != nil {
+				// All "non-terminal" errors are handled by the TSDBBuilder.
+				return state, fmt.Errorf("process record in partition %d at offset %d: %w", rec.Partition, rec.Offset, err)
+			}
+			if !allSamplesProcessed {
+				if lastRec == nil {
+					// The first record was not fully processed, meaning the record before this is the commit point.
+					// We hand-craft the commitRec from the data in the state to re-commit it. On commit the commit's meta is updated
+					// with the new value of LastSeenOffset. This is so the next cycle handled partially processed record properly.
+					commitRec = &kgo.Record{
+						Topic:       state.Commit.Topic,
+						Partition:   state.Commit.Partition,
+						Offset:      state.Commit.At - 1, // Previous commit's offset-1 means we expect the next cycle to start from the commit's offset.
```

Comment:
Could you explain me why - 1? The comment on this line didn't help me.

Reply:
`commitRec` is the record we "commit" to Kafka (similar to [`kgo.Client.CommitRecords`](https://pkg.go.dev/github.com/twmb/franz-go/pkg/kgo#Client.CommitRecords)). But in Kafka, one doesn't commit a record — one commits the offset _after_ the record (i.e. the offset at which next consumption will start).

That was the backstory of this `offset-1`. We create a "fake" record that point at the offset **before** the last commit offset, so the code below committed the offset _after the before_ the last commit offset.

In pseudo-code:

```
lastCommitOffset = 0
recOffset  = lastCommitOffset - 1

commitOffset = recOffset + 1
assert(commitOffset == lastCommitOffset)
kafka.Commit(commitOffset)
```

Reply:
Ok great. You should condense this explanation in a code comment at the line where we do `state.Commit.At - 1` 😉